base/abstractarray.jl

```diff
@@ -600,296 +600,278 @@ get(A::AbstractArray, I::RangeVecIntList, default) = get!(similar(A, typeof(defa
 promote_eltype() = Bottom
 promote_eltype(v1, vs...) = promote_type(eltype(v1), promote_eltype(vs...))
 
-#TODO: ERROR CHECK
-cat(catdim::Integer) = Array(Any, 0)
-
 vcat() = Array(Any, 0)
 hcat() = Array(Any, 0)
-
-## cat: special cases
-hcat{T}(X::T...)         = T[ X[j] for i=1, j=1:length(X) ]
-hcat{T<:Number}(X::T...) = T[ X[j] for i=1, j=1:length(X) ]
-vcat{T}(X::T...)         = T[ X[i] for i=1:length(X) ]
-vcat{T<:Number}(X::T...) = T[ X[i] for i=1:length(X) ]
-
-function vcat(X::Number...)
-    T = promote_typeof(X...)
-    hvcat_fill(Array(T,length(X)), X)
+cat(catdim::Integer) = Array(Any, 0)
+hvcat(rows::(Int...)) = Array(Any, 0)
+
+# auxiliary cat functions
+catlength(catdim, x::AbstractArray) = size(x,catdim)
+catlength(catdim, x) = 1
+
+catrange(catdim, x::AbstractArray) = 1:size(x,catdim)
+catrange(catdim, x) = 1
+
+cattype{S<:AbstractArray}(::Type{S}) = eltype(S)
+cattype(T) = T
+cattypeof{T}(::T) = cattype(T)
+
+promote_cattypeof(x) = cattypeof(x)
+promote_cattypeof(x, X...) = promote_type(cattypeof(x), promote_cattypeof(X...))
+
+catndims(x::AbstractArray) = ndims(x)
+catndims(x) = 1
+
+cat_similar(x::AbstractArray, T, sz) = similar(full(x), T, sz)
```

Comment:
Not sure I follow this, so take these with a big grain of salt:
- Do we really want to require `full` to be implemented for all `AbstractArray`s (or is this used in special circumstances?)
- It would be lovely if this didn't require actual intantiation of `full(x)`, since in the end it's presumably about types and sizes. OTOH, since we don't have `full_type(x), full_size(x)` maybe there isn't a good alternative.